### ISSUE TITLE
Enhance Search bar with autosuggestion and pagination from client based to server based

### ISSUE THREAD

**Issue by Hayat4144** [EddieHubCommunity/BioDrop#9879]
### Is this a unique feature?

- [X] I have checked "open" AND "closed" issues and this is not a duplicate

### Is your feature request related to a problem/unavailable functionality? Please describe.

The current implementation of the search bar employs a client-side pagination strategy, leading to suboptimal performance, especially when dealing with large datasets. One major concern is that the entire document is loaded into the client's memory, resulting in increased latency and potential performance bottlenecks.

The current approach may cause delays in rendering search results, leading to a suboptimal user experience. Users may experience laggy interactions and slower page loads, affecting the usability of the search functionality.

### Proposed Solution

I propose enhancing the current search functionality by adding an autosuggestion feature to the existing search bar and optimizing pagination from client-based to backend-based for improved performance.

### Feature 1: Autosuggestion in Search Bar
Implementing autosuggestion in the search bar can significantly improve the user experience by providing real-time suggestions as users type. This feature can help users refine their search queries, discover relevant content faster, and reduce the likelihood of typos or errors in their input.

### Feature 2: Backend Pagination for Improved Performance
Currently, pagination is handled on the client side, which may lead to suboptimal performance, especially when dealing with a large dataset. By transitioning to backend pagination, we can improve the efficiency of data retrieval and reduce the load on the client. This change can result in faster page load times and a smoother overall user experience.

## Proposed Changes:

 **Autosuggestion in Search Bar**
1. Implement a mechanism for real-time autosuggestions as users type in the search bar.
2. Provide an option for users to easily select and apply suggested queries.

**Backend Pagination:**
1. Refactor the pagination logic to be handled on the server side.
2. Introduce API endpoints that support paginated data retrieval, allowing clients to request specific subsets of data.
3. Update the client-side implementation to make use of the new backend pagination endpoints.

### Screenshots

Here is the some of the screen shot how the ui will looks after the changes.
![Screenshot from 2023-12-05 15-21-07](https://github.com/EddieHubCommunity/BioDrop/assets/100507654/015ec644-2b4e-40a8-9db6-b8573e5e13de)
![Screenshot from 2023-12-05 15-20-29](https://github.com/EddieHubCommunity/BioDrop/assets/100507654/356edc8f-156b-4737-98b1-05fb75bc9cf3)

### Do you want to work on this issue?

Yes

### If "yes" to above, please explain how you would technically implement this

I already solve this problem with mongodb search index and aggregation . if you agree with  this features i will make a pull request.

**Comment by sital002** [comment]
@Hayat4144 Thanks that's a great description on the issue. Can you explain how the auto suggestions work will it make network request to the server on every ```onChange```
event.

@eddiejaoude What are your thought on this.

**Comment by eddiejaoude** [comment]
This sounds great 👍 awesome collaboration 💪 

I have a question, will this also working on a local version of mongo? So dev (local) will be the same as prod (Atlas)? I heard this is now possible with recent changes by mongo

**Comment by Hayat4144** [comment]
I don't think so the autosuggestion is available in mongodb in Os. It is only exist in cloud.

**Comment by eddiejaoude** [comment]
> I don't think so the autosuggestion is available in mongodb in Os. It is only exist in cloud.

How would this affect developing locally? Would it break or return no results or something else?

**Comment by Hayat4144** [comment]
It throw error in local development , but we can prevent it from error. Here is the solution for it.
We use Atlas cloud for production and local mongodb while in development mode. We write logic for handling search functionality based on the app environment in both in frontend and backend  , if it is in development mode we use previous logic and if it is production mode we use atlas search logic same with frontend. we just changed ui nothing more than it.

**Comment by akshat-khosya** [comment]
We can use concept of tries to autocomplete the words that user trying to enter, just like google search.

**Comment by eddiejaoude** [comment]
> We write logic for handling search functionality based on the app environment in both in frontend and backend

I am not keen on this because it will create not only extra code but also when we get bugs and we won't know about them

**Comment by Hayat4144** [comment]
We have to just create the database in cloud add search index.

**Comment by eddiejaoude** [comment]
It looks like Atlas features are available locally now https://www.mongodb.com/blog/post/introducing-local-development-experience-atlas-search-vector-search-atlas-cli

**Comment by Hayat4144** [comment]
if Mongodb Atlas search is available in the local version it's good . we just need mongodbUrl connection which support atlas search either in cloud or locally. Now, Atlas search exist in local database, we should implement this feature.

**Comment by Abhishek-90** [comment]
We can also take it step by step, breaking it up more. 

Firstly, let's create an api for auto suggestions as suggested by @Hayat4144 and test is rigorously on both local and production. If we are satisfied with the results, then we can integrate this api in frontend. 

Coming to frontend, currently we are able to see the results after the user clicks on `search` button. So if we are to implement this feature how will the suggestions appear? Will are be a dropdown of suggestions like `youtube` and `google` or will we directly show the profile like we are currently doing?

**Comment by Hayat4144** [comment]
> We can also take it step by step, breaking it up more.
> 
> Firstly, let's create an api for auto suggestions as suggested by @Hayat4144 and test is rigorously on both local and production. If we are satisfied with the results, then we can integrate this api in frontend.
> 
> Coming to frontend, currently we are able to see the results after the user clicks on `search` button. So if we are to implement this feature how will the suggestions appear? Will are be a dropdown of suggestions like `youtube` and `google` or will we directly show the profile like we are currently doing?

The suggestion will appear as users type in the search bar and a popover will be open just below the search bar. see the below  image 

![Screenshot from 2023-12-05 15-20-29](https://github.com/EddieHubCommunity/BioDrop/assets/100507654/9cc016f9-2a8e-4a2d-8399-f085c09cccb3)

**Comment by Abhishek-90** [comment]
> > We can also take it step by step, breaking it up more.
> > Firstly, let's create an api for auto suggestions as suggested by @Hayat4144 and test is rigorously on both local and production. If we are satisfied with the results, then we can integrate this api in frontend.
> > Coming to frontend, currently we are able to see the results after the user clicks on `search` button. So if we are to implement this feature how will the suggestions appear? Will are be a dropdown of suggestions like `youtube` and `google` or will we directly show the profile like we are currently doing?
> 
> The suggestion will appear as users type in the search bar and a popover will be open just below the search bar. see the below image
> 
> ![Screenshot from 2023-12-05 15-20-29](https://private-user-images.githubusercontent.com/100507654/292534097-9cc016f9-2a8e-4a2d-8399-f085c09cccb3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMzMTMwODksIm5iZiI6MTcwMzMxMjc4OSwicGF0aCI6Ii8xMDA1MDc2NTQvMjkyNTM0MDk3LTljYzAxNmY5LTJhOGUtNGEyZC04Mzk5LWYwODVjMDljY2NiMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIyM1QwNjI2MjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iM2Y1YTg3NmJjMjE5NDJiOGNlMDU5ZjFiOTJkZWVkNjdkMmI3OGZjZGM4MDBmZmY2NjNkNDI4MWYxY2M0MzYwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.vCe_UF7SuO19M5k_Ks_vgSR46EHSi1yMfYLN4654uso)

Ok and when we click on suggestions, we will be redirected to view the full information of the user, correct? 

I have one other suggestion, instead of adding a new feature to get suggestions, how about we remove the search button all-together. What we can do is, merge this suggestion and search feature. As mentioned by @Hayat4144 we will use `debounce` pattern. When user types in search we will hit the api without clicking on search button. and directly populate the data to view the profiles. The benefit of this approuch would be: 

1. Less UI changes, we don't need to create a new suggestion dropdown list to show the suggestion, we will be utilizing current view only. 
2. No need to create a new api end-point, rather makes changes to existing to search api endpoint.

**Comment by Hayat4144** [comment]
The existing search functionality is something like the below steps:- 

1. user give their input either tag or name and click on the button.
2. it will it hit the `/api/search` endpoint where the whole database users is fetched and then some logic to retrive the similar users based on the tag or name. This is silly thing. we don't need to fetch every user and then extract the related user.
3. the response will be send back with list of users .
4. it will shown in the ui.

Here is what i suggest 

1. create one more endpoint for the autosuggestion , when the start typing we will hit this endpoint  it will fetch the users based on search query and send it back to the client.
2. the autosuggestion api response will be shown just below the search bar with popover .
3. In third step we have two choice when the user click on one of the result we will either fetch by `name or tags. `
4. Response will be shown on the page

Using either `Tag` or `Name ` have advantage and disadvantage
if we use tag for fetching the final users , we will fetch the all users have similar tag but 90% the user details we click on will be not shown in the top,we have to add some sort of scoring in mongodb similar to Rank we do int postgres Full text search.

if use name we will fetch the user having similar name, but not with similar tags.
Currently , I implemented with tags.
